### PR TITLE
[stdlib_math] Add arange function.

### PR DESCRIPTION
**Pull request by zoziha**
#### Description
`seq` returns evenly spaced values within a given interval.
- [x] Add `seq` function. (see [numpy/arange](https://numpy.org/devdocs/reference/generated/numpy.arange.html#numpy.arange))  
(the API name `seq` needs to be discussed, `arange`? I think `seq` is better, it is shorter.) 
(aready to `arange`🔰)
#### Tasks
- [x] `by <= 0` condition. (see https://github.com/fortran-lang/stdlib/pull/480#discussion_r680382134)
- [x] put it in `stdlib_math` or `stdlib_linalg`: `stdlib_math` is better.
- [x] `floor` or `ceiling` function: I think `floor` is better. (see https://github.com/fortran-lang/stdlib/pull/480#discussion_r680382361)
- [x] `by` value is a very small amount. (Need to discuss) (see https://github.com/fortran-lang/stdlib/pull/480#discussion_r681494363) (let users judge😏)

#### More routines to do, not this PR
see #476.

### CODE REVIEW

File: doc/specs/stdlib_math.md [hunk 1]
```diff
@@ -275,3 +275,61 @@ program demo_logspace_rstart_cbase
...
+
+`end`: Shall be an `integer/real` scalar.
+This is an `intent(in)` and `optional` argument.
+
+`by`: Shall be an `integer/real` scalar and large than `0`.
+This is an `intent(in)` and `optional` argument.  
+
+Warning:  
+If `by = 0`, the `by` argument will be corrected to `1/1.0` by the internal process of the `seq` function.   
+If `by < 0`, the `by` argument will be corrected to `abs(by)` by the internal process of the `seq` function.
```
Comment (zoziha) [inline doc/specs/stdlib_math.md:309]: `by <= 0` condition. Is this way ok?

File: src/stdlib_math_seq.fypp [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        integer :: i
+
+        start_ = merge(start, 1.0_${k1}$, present(end))
+        end_   = merge(end,   start,      present(end))
+        by_    = sign(merge(merge(by, 1.0_${k1}$, by /= 0.0_${k1}$), &
+                 1.0_${k1}$, present(by)), end_ - start_)
+
+        allocate(result(floor((end_ - start_)/by_) + 1))
+            #!TODO: ??
+            #! floor or ceiling, floor is better.
```
Comment (zoziha) [inline src/stdlib_math_seq.fypp:27]: When `ceiling`:
```fortran
seq(1.0, 3.0, 1.0) = [1.0, 2.0]
seq(1.0, 3.1, 1.0) = [1.0, 2.0, 3.0]
```
When `floor`:
```fortran
seq(1.0, 3.0, 1.0) = [1.0, 2.0, 3.0]
```
So `floor` is better!

**Comment by awvwgk** [comment]
This looks like a useful addition, but the name `seq` seems rather short. Why not use `arange` here? `sequence` would also be an option, but this is already an attribute for derived types used for memory alignment or creating non-extendible types.

**Comment by zoziha** [comment]
In fact, the name `seq` is derived from the `seq` function in the `R` language, taking into account the experience in `fortran`:
```fortran
v = seq(10.0_dp, 100.0_dp, 10.0_dp)    !! It is shorter.
v = (1.0_dp, 1.0_dp)*seq((10.0_dp, 100.0_dp, 10.0_dp)
---
v = arange(10.0_dp, 100.0_dp, 10.0_dp) !! It is OK.
v = (1.0_dp, 1.0_dp)*arange((10.0_dp, 100.0_dp, 10.0_dp)
```
`seq` has three arguments, I hope this API name can be shorter, but I also like `arange` very much, next time I will consider replacing `seq` with `arange`💖.

File: src/stdlib_math_arange.fypp [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        
+        ${t1}$ :: start_, end_, by_
+        integer :: i
+
+        start_ = merge(start, 1.0_${k1}$, present(end))
+        end_   = merge(end,   start,      present(end))
+        by_    = sign(merge(merge(by, 1.0_${k1}$, by /= 0.0_${k1}$), &
+                 1.0_${k1}$, present(by)), end_ - start_)
+
+        allocate(result(floor((end_ - start_)/by_) + 1))
```
Comment (zoziha) [inline src/stdlib_math_arange.fypp:25]:
```fortran
by /= 0.0_${k1}$
```
When the `by` value is a very small amount, here becomes dangerous. Do we need to set a limit for it?

File: src/stdlib_math_arange.fypp [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+        result = [(start_ + (i - 1)*by_, i=1, size(result), 1)]
```
Comment (zoziha) [inline src/stdlib_math_arange.fypp:27]:
```fortran
i = size(result)    !! We know: huge(integer :: i) = 2147483647
```
So `arange` only can create a vector with a maximum length 2147483647.

**Comment by Jim-215-Fisher** [comment]
I will review this one.

File: src/stdlib_math_arange.fypp (hunk 4 above)
Comment (Jim-215-Fisher) [reply src/stdlib_math_arange.fypp:27]: This doesn't sound right for the real values. How about arange(1.0,6.0,2.0) which should give us [1.0, 3.5, 6.0]. Here it gives us [1.0, 3.0, 5.0].

**Comment by zoziha** [comment]
> This doesn't sound right for the real values. How about arange(1.0,6.0,2.0) which should give us [1.0, 3.5, 6.0]. Here it gives us [1.0, 3.0, 5.0].

Thanks for your review! @Jim-215-Fisher  `arange` is different with `linspace`, I think you may be talking about the function of [`linspace`](https://github.com/fortran-lang/stdlib/blob/master/doc/specs/stdlib_math.md#linspace---create-a-linearly-spaced-rank-one-array).
(see numpy: [arange](https://numpy.org/devdocs/reference/generated/numpy.arange.html#numpy.arange) and [linspace](https://numpy.org/devdocs/reference/generated/numpy.linspace.html))

**Comment by Jim-215-Fisher** [comment]
> > This doesn't sound right for the real values. How about arange(1.0,6.0,2.0) which should give us [1.0, 3.5, 6.0]. Here it gives us [1.0, 3.0, 5.0].
> 
> Thanks for your review! @Jim-215-Fisher `arange` is different with `linspace`, I think you may be talking about the function of [`stdlib_math(module):linspace(interface)`](https://github.com/fortran-lang/stdlib/blob/master/doc/specs/stdlib_math.md#linspace---create-a-linearly-spaced-rank-one-array).
> (see numpy: [arange](https://numpy.org/devdocs/reference/generated/numpy.arange.html#numpy.arange) and [linspace](https://numpy.org/devdocs/reference/generated/numpy.linspace.html))

After reading numpy's documents, I understand that `arange` has a fixed spacing while `linspace` has fixed number of samples. The fixed number of samples give evenly spaced results. The fixed spacing in `arange` may give "evenly" spaced results without end point. My suggestion would be to change the wording of description for the function from "evenly spaced values" to "fix spaced values".

Another question is whether this function is elemental. If it is, why not change "pure" to "elemental"? In numpy, 'arange' is suitable to array arguments.

**Comment by zoziha** [comment]
> Another question is whether this function is elemental. If it is, why not change "pure" to "elemental"? In numpy, 'arange' is suitable to array arguments.

Thanks, `fixed spaced values` is indeed better. 

This function is not elemental. It does not seem to meet the definition of `elemental` function.
I have not used numpy, but it seems that `numpy.arange` does not state that it supports array arguments. 

Do you mean the return value is an array (not only a vector)? In addition, it is difficult for the existing `fortran` syntax to implement a multi-dimensional array generic interface. (see https://github.com/j3-fortran/fortran_proposals/issues/153)

**Comment by Jim-215-Fisher** [comment]
> > Another question is whether this function is elemental. If it is, why not change "pure" to "elemental"? In numpy, 'arange' is suitable to array arguments.
> 
> Thanks, `fixed spaced values` is indeed better.
> 
> This function is not elemental. It does not seem to meet the definition of `elemental` function.
> I have not used numpy, but it seems that `numpy.arange` does not state that it supports array arguments.
> 
> Do you mean the return value is an array (not only a vector)? In addition, it is difficult for the existing `fortran` syntax to implement a multi-dimensional array generic interface. (see [j3-fortran/fortran_proposals#153](https://github.com/j3-fortran/fortran_proposals/issues/153))

You are right. numpy.arange does not support array argument. Never mind.

**Review by Jim-215-Fisher — APPROVED** [review]

File: doc/specs/stdlib_math.md [hunk 5]
```diff
@@ -275,3 +275,75 @@ program demo_logspace_rstart_cbase
...
+This is an `intent(in)` and `optional` argument.   
+The default `by` value is `1`.
+
+#### Warning
+If `by = 0`, the `by` argument will be corrected to `1/1.0` by the internal process of the `arange` function.   
+If `by < 0`, the `by` argument will be corrected to `abs(by)` by the internal process of the `arange` function. 
+
+### Return value
+
+Return a vector of fixed spaced spaced values.
```
Comment (milancurcic) [inline doc/specs/stdlib_math.md:322]:
```suggestion
Returns an array of fixed-space values.
```

File: doc/specs/stdlib_math.md [hunk 6]
```diff
@@ -275,3 +275,75 @@ program demo_logspace_rstart_cbase
...
+
+Experimental
+
+### Class
+
+Pure function.
+
+### Description
+
+Create a vector of the `integer/real` type with given fixed spaced values within a given interval.
```
Comment (milancurcic) [inline doc/specs/stdlib_math.md:290]:
```suggestion
Creates an array of the `integer/real` type with fixed-spaced values of given spacing, within a given interval.
```

File: doc/specs/stdlib_math.md [hunk 7]
```diff
@@ -275,3 +275,75 @@ program demo_logspace_rstart_cbase
...
+
+#### Note
+
+Because of the `i` (`huge(integer :: i) = 2147483647`) index inside the `arange` function , the dimensional maximum length of array created by the `arange` function is `2147483647`.
```
Comment (milancurcic) [inline doc/specs/stdlib_math.md:294]: Is there a use case to create larger arrays? I can't think of one, but if there is, we could use `int64` to allow larger arrays.

File: doc/specs/stdlib_math.md [hunk 8]
```diff
@@ -275,3 +275,75 @@ program demo_logspace_rstart_cbase
...
+
+Create a vector of the `integer/real` type with given fixed spaced values within a given interval.
+
+#### Note
+
+Because of the `i` (`huge(integer :: i) = 2147483647`) index inside the `arange` function , the dimensional maximum length of array created by the `arange` function is `2147483647`. 
+
+### Syntax
+
+`result = [[stdlib_math(module):arange(interface)]](start [, end, by])`
```
Comment (milancurcic) [inline doc/specs/stdlib_math.md:298]: `by` for the third argument sounds odd to me. I'm trying to think who to say this in words: Arange from start to end by...? It doesn't sound right. 

Why not call this `increment`, or just `incr` if `increment` is too long? Or `step`, like in numpy.arange?

File: src/stdlib_math_arange.fypp [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+        ${t1}$, intent(in), optional :: end, by
+        ${t1}$, allocatable :: result(:)
+        
+        ${t1}$ :: start_, end_, by_
+        integer :: i
+
+        start_ = merge(start, 1.0_${k1}$, present(end))
+        end_   = merge(end,   start,      present(end))
+        by_    = sign(merge(merge(by, 1.0_${k1}$, by /= 0.0_${k1}$), &
+                 1.0_${k1}$, present(by)), end_ - start_)
```
Comment (milancurcic) [inline src/stdlib_math_arange.fypp:21]: If `end` or `by` are not present, one of these lines will be standard non-conforming and possibly segfault. We should use `optval` from `stdlib_optval` instead.

File: src/stdlib_math_arange.fypp [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+        ${t1}$, intent(in), optional :: end, by
+        ${t1}$, allocatable :: result(:)
+        
+        ${t1}$ :: start_, end_, by_
+        ${t1}$ :: i
+
+        start_ = merge(start, 1_${k1}$, present(end))
+        end_   = merge(end,   start,    present(end))
+        by_    = sign(merge(merge(by, 1_${k1}$, by /= 0_${k1}$), &
+                 1_${k1}$, present(by) ), end_ - start_)
```
Comment (milancurcic) [inline src/stdlib_math_arange.fypp:45]: Ditto here, `optval` instead of `merge`.

**Review by milancurcic — CHANGES_REQUESTED** [review]
Thank you for this PR! I think the API is mostly good, I gave some naming suggestions to use instead of `by`, and I think there is an issue with using `present()` inside `merge()`, so I suggested using `optval()` instead.

File: doc/specs/stdlib_math.md (hunk 6 above)
Comment (zoziha) [reply doc/specs/stdlib_math.md:290]: In fact, what `arange` returns is a one-dimensional `vector`, the word `array` may be ambiguous?

File: doc/specs/stdlib_math.md (hunk 7 above)
Comment (zoziha) [reply doc/specs/stdlib_math.md:294]: I think the length of `2147483647` as an array dimension is very sufficient. `int64` seems unnecessary, but I may subconsciously explain this in the document. Does it need to be removed?

File: doc/specs/stdlib_math.md (hunk 8 above)
Comment (zoziha) [reply doc/specs/stdlib_math.md:298]: OK, `step`.😏

File: doc/specs/stdlib_math.md (hunk 6 above)
Comment (milancurcic) [reply doc/specs/stdlib_math.md:290]: To me it's the other way around--I think of vector as a mathematical concept and array as a programming one. I think array is also a more Fortrannic word. But it's not a big deal, just a suggestion, pick your favorite, either works well 👍

File: doc/specs/stdlib_math.md (hunk 7 above)
Comment (milancurcic) [reply doc/specs/stdlib_math.md:294]: Okay, great, I was just wondering. I think it's okay that it stays int32. Also, the note is helpful IMO.

File: src/stdlib_math_arange.fypp (hunk 9 above)
Comment (zoziha) [reply src/stdlib_math_arange.fypp:21]: I tested it (see https://github.com/fortran-lang/stdlib/issues/448#issuecomment-893110485), it is indeed what you said, I will change this kind of `merge` to `optval`.

File: doc/specs/stdlib_math.md (hunk 6 above)
Comment (zoziha) [reply doc/specs/stdlib_math.md:290]: I think it can be taken as `one-dimensional array`

**Review by milancurcic — APPROVED** [review]

**Review by awvwgk — APPROVED** [review]
Thanks for sharing, looks good to me.

**Comment by milancurcic** [comment]
With 3 approvals, I'll go ahead and merge. Thank you @zoziha and reviewers!

**Pull request merged**